apps/emqx/src/emqx_router_helper.erl

```diff
@@ -188,6 +230,228 @@ code_change(_OldVsn, State, _Extra) ->
 %% Internal functions
 %%--------------------------------------------------------------------
 
+handle_membership_event({node, down, Node}, State) ->
+    _ = record_node_down(Node),
+    State;
+handle_membership_event({node, leaving, Node}, State) ->
+    _ = schedule_purge_left(Node),
+    _ = record_node_left(Node),
+    State;
+handle_membership_event({node, up, Node}, State) ->
+    _ = record_node_alive(Node),
+    State;
+handle_membership_event({mnesia, up, Node}, State = #{last_membership := Membership}) ->
+    State#{last_membership := lists:usort([Node | Membership])};
+handle_membership_event(_Event, State) ->
+    State.
+
+record_node_down(Node) ->
+    NS = #ns{
+        node = Node,
+        status = down,
+        since = erlang:monotonic_time(millisecond)
+    },
+    case ets:lookup(?TAB_STATUS, Node) of
+        [#ns{status = left}] ->
+            %% Node is still marked as left, keep it that way.
+            false;
+        [#ns{status = down}] ->
+            %% Duplicate.
+            ets:insert(?TAB_STATUS, NS);
+        [] ->
+            ets:insert(?TAB_STATUS, NS)
+    end.
+
+record_node_left(Node) ->
+    NS = #ns{
+        node = Node,
+        status = left,
+        since = erlang:monotonic_time(millisecond)
+    },
+    ets:insert(?TAB_STATUS, NS).
+
+record_node_alive(Node) ->
+    forget_node(Node).
```

Comment:
write a warning message ?

Reply:
In a similar vein, there's https://github.com/emqx/mria/blob/603bea2481917041e502952e834ccc53ddc876f0/src/mria_membership.erl#L324 already.